### PR TITLE
Origin/dev/adf

### PR DESCRIPTION
**Pull request by afeinstein20**
I deleted the old pull request in exchange for this one. This one has 2 different ways to trace and extract the orders (they give slightly different solutions) and has a function to load in the wavelength solution. This pull request also has minor changes to setup files.

### CODE REVIEW

File: eureka/S3_data_reduction/niriss.py [hunk 1]
```diff
@@ -1,2 +1,435 @@
...
+    gcenters_1 = rm_outliers(gcenters_1)
+    gcenters_2 = rm_outliers(gcenters_2)
+    x = np.arange(0,len(gcenters_1),1)
+
+    fit1 = clean_and_fit(x, x[x>800],
+                         f_centers, gcenters_1[x>800])
+    fit2 = clean_and_fit(x, x[(x>800) & (x<1800)],
+                         f_centers, gcenters_2[(x>800) & (x<1800)])
+    
+    if isplots:
```
Comment (kevin218) [inline eureka/S3_data_reduction/niriss.py:259]: You know the drill...

File: eureka/S3_data_reduction/niriss.py [hunk 2]
```diff
@@ -1,2 +1,435 @@
...
+        else:
+            ytot = np.append(f277_peaks[:,1][:cutoff], y)
+        
+        # Fits a 4th degree polynomiall
+        poly= np.polyfit(xtot, ytot, deg=4)
+        fit = np.poly1d(poly)
+            
+        avg[:,ind] = fit(x)
+
+    if isplots:
```
Comment (kevin218) [inline eureka/S3_data_reduction/niriss.py:359]: ditto

File: eureka/S3_data_reduction/niriss.py [hunk 3]
```diff
@@ -1,2 +1,435 @@
...
+       An option to plot the data and intermediate steps to
+       retrieve the mask per each order. Default is False.  
+    """
+    perc  = np.nanmax(data.data, axis=0)
+
+    # creates data img mask
+    z,g = image_filtering(perc)
+    
+    if isplots:
+        plt.close()
```
Comment (kevin218) [inline eureka/S3_data_reduction/niriss.py:396]: this close statement confuses me

**Review by kevin218 — CHANGES_REQUESTED** [review]
It may be easier to chat about some of my comments.  It can be difficult to track exactly what's going on when just looking at the code.

File: eureka/S3_data_reduction/background.py [hunk 4]
```diff
@@ -294,23 +297,37 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
...
+
                     # Mask data point if > threshold
                     if stdevs[loc] > threshold:
-                        mask[j,goodxvals[loc]] = 0
+                        bgmask[j,goodxvals[loc]] = np.nan#0
                     else:
                         nobadpixels = True      #exit while loop
 
+
+                    if isplots == True:
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:322]: updated in the recent push

File: eureka/S3_data_reduction/niriss.py [hunk 5]
```diff
@@ -1,2 +1,435 @@
...
+        inds = np.where(mask[:,i]==True)[0]
+        if len(inds) > 1:
+            new_mask[inds[1]:inds[-2], i] = True
+            mid[i] = np.array([i, (inds[1]+inds[-2])/2])
+
+    q = ((mid[:,0]<420) & (mid[:,1]>0) & (mid[:,0] > 0))
+
+    data.f277_img = new_mask
+
+    if isplots:
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/niriss.py]: updated in the recent push

File: eureka/S3_data_reduction/niriss.py (hunk 1 above)
Comment (afeinstein20) [reply eureka/S3_data_reduction/niriss.py]: updated in the recent push

File: eureka/S3_data_reduction/niriss.py (hunk 2 above)
Comment (afeinstein20) [reply eureka/S3_data_reduction/niriss.py]: updated in the recent push

File: eureka/S3_data_reduction/niriss.py (hunk 3 above)
Comment (afeinstein20) [reply eureka/S3_data_reduction/niriss.py]: I think I was having trouble with this overplotting in my notebook with another figure. But by itself it works fine. I've removed the close statement and updated the isplots like the other ones (including a isplots>=5)

File: eureka/S3_data_reduction/background.py [hunk 6]
```diff
@@ -246,42 +248,43 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
     # Initiate background image with zeros
     ny, nx  = np.shape(dataim)
     bg      = np.zeros((ny,nx))
-    mask2   = mask*bgmask
-    if deg < 0:
+
+    if deg <= 0:
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:252]: this has been trashed

File: eureka/S3_data_reduction/background.py [hunk 7]
```diff
@@ -294,23 +297,37 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
...
-                    stdres      = np.std(residuals[ind])
+                    ind = np.arange(0,len(residuals),1)
+                    ind = np.delete(ind, loc)
+                    stdres = np.std(residuals[ind])
+                    
                     if stdres == 0:
                         stdres = np.inf
                     # Calculate number of sigma from the model
                     stdevs    = np.abs(residuals) / stdres
+                    #print(stdevs)
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:313]: this has been trashed

File: eureka/S3_data_reduction/background.py [hunk 8]
```diff
@@ -294,23 +297,37 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
...
+
                     # Mask data point if > threshold
                     if stdevs[loc] > threshold:
-                        mask[j,goodxvals[loc]] = 0
+                        bgmask[j,goodxvals[loc]] = np.nan#0
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:317]: this has been trashed

File: eureka/S3_data_reduction/background.py [hunk 9]
```diff
@@ -294,23 +297,37 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
...
                     else:
                         nobadpixels = True      #exit while loop
 
+
+                    if isplots == True:
+                        plt.figure(3601)
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:323]: this has been trashed

File: eureka/S3_data_reduction/background.py [hunk 10]
```diff
@@ -325,4 +342,92 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
         mask    = (mask.T)
         bgmask  = (bgmask.T)
 
-    return bg, mask #,variance
+    return bg*bgmask#, mask #,variance
```
[the remark this answers is not in the document]
Comment (afeinstein20) [reply eureka/S3_data_reduction/background.py:345]: this has been trashed

**Comment by afeinstein20** [comment]
I've gone through and (1) updated changes to the `isplots` argument and resolved all of those comments and (2) resolved all comments addressing the `background.py` file, since those have all been scrapped.

**Comment by taylorbell57** [comment]
Hi Adina, I just looked through the code and it seems all your changes to background.py that Kevin commented on are still present - are there commits you haven't pushed to GitHub yet?

**Comment by afeinstein20** [comment]
Hmm, it should have been changed in the last push I did yesterday. Is it removed now?

**Comment by taylorbell57** [comment]
No, I still see all the changes - look in the Files Changed tab on the pull request here on GitHub

**Comment by afeinstein20** [comment]
If I go to the Files Changed >> background.py (under the last push `Update background.py`) I see the entire `fitbg3` function is now gone, with the exception of some docstrings

**Comment by taylorbell57** [comment]
`fitbg3` is indeed gone, but Kevin's comments on your other edits to the background.py file are still relevant (e.g. see the attached comment which is still relevant) 
Screen Shot 2022-01-25 at 12 14 22

**Comment by afeinstein20** [comment]
Ah, oops!! Yes, I've gone and fixed those now

**Review by kevin218 — APPROVED** [review]

**Pull request merged**